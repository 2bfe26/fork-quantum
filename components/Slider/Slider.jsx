import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import RcSlider from 'rc-slider';
import 'rc-slider/assets/index.css';
import Colors from '../Colors';
import Tooltip from '../Tooltip';
import valueValidator from './valueValidator';

const sliderStyle = {
  handleStyle: {
    borderWidth: 0,
    height: 20,
    marginLeft: -9,
    marginTop: -6,
    width: 20,
  },
  trackStyle: {
    height: 8,
  },
  railStyle: {
    height: 8,
  },
  marksStyle: {
    common: {
      fontSize: 16,
      marginLeft: 0,
      width: 'auto',
    },
    min: {
      left: 0,
    },
    max: {
      left: 'unset',
      right: 0,
    },
  },
};

const customStyle = css`
  .rc-slider-handle {
    background-color: ${Colors.BLUE[500]};
  }

  .rc-slider-rail {
    background-color: ${Colors.BLUE[200]};
  }

  .rc-slider-track {
    background-color: ${Colors.BLUE[50]};
  }

  .rc-slider-handle:active,
  .rc-slider-handle:focus {
    border: none;
    box-shadow: 0 2px 6px 0 ${Colors.BLUE[50]};
  }

  &.rc-slider-disabled {
    background: none;

    .rc-slider-handle {
      background-color: ${Colors.BLACK[400]};
    }

    .rc-slider-rail {
      background-color: ${Colors.BLACK[100]};
    }

    .rc-slider-track {
      background-color: ${Colors.BLACK[200]};
    }
  }
`;

const StyledSlider = styled(RcSlider)`
  ${customStyle}
`;
const StyledRange = styled(RcSlider.Range)`
  ${customStyle}
`;

const StyledTooltip = styled(Tooltip)`
  width: 100%;

  > div:first-child {
    top: -41px;
    left: ${({ value, min, max }) => {
      let half = value;

      if (typeof value === 'object') {
        const { from, to } = value;
        half = (to + from) / 2;
      }

      return ((half - min) / (max - min)) * 100;
    }}%;
  }
`;

class Slider extends React.Component {
  constructor(props) {
    super(props);

    this.state = { visible: null };
  }

  handleChange = value => {
    const { onChange } = this.props;

    if (Array.isArray(value)) {
      const [from, to] = value;
      onChange({ from, to });
    } else {
      onChange(value);
    }
  };

  handleMouseDown = () => {
    this.setState({ visible: true });
  };

  handleMouseUp = () => {
    this.setState({ visible: false });
  };

  render() {
    const {
      handleChange,
      handleMouseDown,
      handleMouseUp,
      props,
      state: { visible },
    } = this;

    const { value, tipFormatter, minMaxFormatter, min, max } = props;
    const { from, to } = value;
    const { handleStyle, trackStyle, marksStyle, railStyle } = sliderStyle;
    const marks = {
      [min]: {
        style: {
          ...marksStyle.common,
          ...marksStyle.min,
        },
        label: minMaxFormatter(min),
      },
      [max]: {
        style: {
          ...marksStyle.common,
          ...marksStyle.max,
        },
        label: minMaxFormatter(max),
      },
    };

    return (
      <StyledTooltip
        {...props}
        text={String(tipFormatter(value))}
        value={value}
        visible={visible}
      >
        {typeof value === 'object' ? (
          <StyledRange
            {...props}
            allowCross={false}
            dotStyle={{ display: 'none' }}
            handleStyle={[handleStyle, handleStyle]}
            max={max}
            min={min}
            marks={marks}
            pushable
            onChange={handleChange}
            onBeforeChange={handleMouseDown}
            onAfterChange={handleMouseUp}
            railStyle={railStyle}
            trackStyle={[trackStyle, trackStyle]}
            value={[from, to]}
          />
        ) : (
          <StyledSlider
            {...props}
            dotStyle={{ display: 'none' }}
            handleStyle={handleStyle}
            max={max}
            marks={marks}
            min={min}
            onChange={handleChange}
            onBeforeChange={handleMouseDown}
            onAfterChange={handleMouseUp}
            railStyle={railStyle}
            trackStyle={trackStyle}
            value={value}
          />
        )}
      </StyledTooltip>
    );
  }
}

StyledRange.displayName = 'RcRange';
StyledSlider.displayName = 'RcSlider';

Slider.defaultProps = {
  max: 100,
  min: 0,
  value: 50,
  disabled: false,
  onChange: () => {},
  tipFormatter: value =>
    typeof value === 'object' ? `${value.from} to ${value.to}` : value,
  minMaxFormatter: value => value,
};

Slider.propTypes = {
  max: PropTypes.number,
  min: PropTypes.number,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  /** Slider will pass its value to tipFormatter, display its value in Tooltip, and hide Tooltip when return value is null. */
  tipFormatter: PropTypes.func,
  minMaxFormatter: PropTypes.func,
  value: valueValidator,
};

export default Slider;
