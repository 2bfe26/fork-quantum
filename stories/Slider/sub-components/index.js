import ControlledSlider from '../ControlledSlider';
import ControlledTwoHandles from './ControlledTwoHandles';
import Basic from '../Basic';
import TipFormatter from '../TipFormatter';
import MinMaxFormatter from '../MinMaxFormatter';

export {
  ControlledSlider,
  ControlledTwoHandles,
  Basic,
  TipFormatter,
  MinMaxFormatter,
};
