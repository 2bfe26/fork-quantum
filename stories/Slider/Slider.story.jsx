import React from 'react';
import { storiesOf } from '@storybook/react';
import {
  AutoExample,
  Tab,
  Example,
  Title,
} from '@catho-private/quantum-storybook-ui';
import Slider from '../../components/Slider';

import { ControlledSlider, ControlledTwoHandles } from './sub-components';

const twoHandles = <Slider value={{ from: 0, to: 100 }} />;

const tipFormatter = <Slider tipFormatter={value => `R$ ${value}`} />;
const tipFormatterWithTwoHandles = (
  <Slider
    value={{ from: 10, to: 80 }}
    tipFormatter={({ from, to }) => `From ${from} to ${to} functionaries`}
  />
);

const tabExample = (
  <Tab title="Examples">
    <Title>With two handles</Title>
    <Example component={twoHandles} />
    <Title>tipFormatter</Title>
    <p>tipFormatter is a prop to format the tooltip value.</p>
    <Example
      component={tipFormatter}
      code="<Slider tipFormatter={value => `R$ ${value}`} />" // eslint-disable-line
    />
    <Example
      component={tipFormatterWithTwoHandles}
      code="<Slider tipFormatter={({from, to}) => `From ${from} to ${to} functionaries`} />" // eslint-disable-line
    />
    <Title>Controlled</Title>
    <Example component={<ControlledSlider />} code={ControlledSlider.code} />
    <Title>Controlled with two handles</Title>
    <Example
      component={<ControlledTwoHandles />}
      code={ControlledTwoHandles.code}
    />
  </Tab>
);

storiesOf('12. Slider', module).add('Slider', () => (
  <AutoExample component={Slider} additionalTabs={tabExample} />
));
