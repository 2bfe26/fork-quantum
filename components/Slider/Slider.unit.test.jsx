import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Slider from './Slider';
import 'jest-styled-components';

describe('<Slider />', () => {
  describe('Snapshots', () => {
    it('should match the snapshot when have just one handle', () => {
      const range = mount(<Slider value={10} />);
      expect(toJson(range)).toMatchSnapshot();
    });

    it('should match the snapshot when have two handles', () => {
      const range = mount(<Slider value={{ from: 10, to: 30 }} />);
      expect(toJson(range)).toMatchSnapshot();
    });

    it('should match the snapshot when <Slider /> is disabled', () => {
      const range = mount(<Slider disabled />);
      expect(toJson(range)).toMatchSnapshot();
    });
  });

  describe('Tooltip position', () => {
    it('should align <Tooltip /> correctly when have just one handle', () => {
      const range = mount(<Slider value={50} />);
      const tooltip = range.find('Tip');

      expect(tooltip).toHaveStyleRule('left', '50%');
    });

    it('shuold align <Tooltip /> correctly when have two handles', () => {
      const range = mount(<Slider value={{ from: 0, to: 100 }} />);
      const tooltip = range.find('Tip');

      expect(tooltip).toHaveStyleRule('left', '50%');
    });
  });

  describe('tipFormatter', () => {
    it('should format Tooltip text and display it correctly when have just one handle', () => {
      const range = mount(
        <Slider value={10} tipFormatter={value => `R$ ${value}`} />,
      );
      const tooltip = range.find('Tip');

      expect(tooltip.text()).toBe('R$ 10');
    });

    it('should format Tooltip text and display it correctly when have two handles', () => {
      const range = mount(
        <Slider
          value={{ from: 10, to: 40 }}
          tipFormatter={({ from, to }) => `${from}km to ${to}km`}
        />,
      );
      const tooltip = range.find('Tip');

      expect(tooltip.text()).toBe('10km to 40km');
    });
  });

  describe('min and max props', () => {
    it('should pass min prop to inner rc-slider component', () => {
      const min = 10;

      const range = mount(<Slider min={min} value={{ from: 10, to: 100 }} />);
      const slider = mount(<Slider min={min} />);

      const rcSlider = slider.find('rcSlider');
      const rcRange = range.find('rcRange');

      expect(rcSlider.props().min).toBe(min);
      expect(rcRange.props().min).toBe(min);
    });

    it('should pass max prop to inner rc-slider component', () => {
      const max = 100;

      const range = mount(<Slider max={max} value={{ from: 0, to: 100 }} />);
      const slider = mount(<Slider max={max} />);

      const rcSlider = slider.find('rcSlider');
      const rcRange = range.find('rcRange');

      expect(rcSlider.props().max).toBe(max);
      expect(rcRange.props().max).toBe(max);
    });
  });

  describe('value prop', () => {
    it('should convert value object to array and pass to inner rc-slider component', () => {
      const range = mount(<Slider value={{ from: 0, to: 50 }} />);
      const slider = range.find('rcRange');

      expect(slider.props().value).toEqual([0, 50]);
    });
  });
});
