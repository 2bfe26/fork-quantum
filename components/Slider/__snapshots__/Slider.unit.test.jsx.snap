// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider component  should match the snapshot 1`] = `
<div
  className="rc-slider Slider__StyledSlider-hqytAK cTstbT"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  style={undefined}
>
  <div
    className="rc-slider-rail"
    style={Object {}}
  />
  <div
    className="rc-slider-track"
    style={
      Object {
        "left": "0%",
        "width": "0%",
      }
    }
  />
  <div
    className="rc-slider-step"
  />
  <div
    aria-disabled={false}
    aria-valuemax={50}
    aria-valuemin={0}
    aria-valuenow={0}
    className="rc-slider-handle"
    role="slider"
    style={
      Object {
        "left": "0%",
      }
    }
    tabIndex={0}
  />
  <div
    className="rc-slider-mark"
  />
</div>
`;
